src/ray/util/logging.cc

```diff
@@ -0,0 +1,253 @@
+#include <cstdlib>
+#include <iostream>
+
+#include "ray/util/logging.h"
+
+#ifdef RAY_USE_GLOG
+#include "glog/logging.h"
+#elif RAY_USE_LOG4CPLUS
+#include <log4cplus/consoleappender.h>
+#include <log4cplus/fileappender.h>
+#include <log4cplus/layout.h>
+#include <log4cplus/logger.h>
+#include <log4cplus/loggingmacros.h>
+#endif
+
+namespace ray {
+
+// This is the default implementation of ray log,
+// which is independent of any libs.
+class CerrLog {
+ public:
+  CerrLog(int severity)  // NOLINT(runtime/explicit)
+      : severity_(severity),
+        has_logged_(false) {}
+
+  virtual ~CerrLog() {
+    if (has_logged_) {
+      std::cerr << std::endl;
+    }
+    if (severity_ == RAY_FATAL) {
+      PrintBackTrace();
+      std::abort();
+    }
+  }
+
+  std::ostream &Stream() {
+    has_logged_ = true;
+    return std::cerr;
+  }
+
+  template <class T>
+  CerrLog &operator<<(const T &t) {
+    if (severity_ != RAY_DEBUG) {
+      has_logged_ = true;
+      std::cerr << t;
+    }
+    return *this;
+  }
+
+ protected:
+  const int severity_;
+  bool has_logged_;
+
+  void PrintBackTrace() {
+#if defined(_EXECINFO_H) || !defined(_WIN32)
+    void *buffer[255];
+    const int calls = backtrace(buffer, sizeof(buffer) / sizeof(void *));
+    backtrace_symbols_fd(buffer, calls, 1);
+#endif
+  }
+};
+
+// This is a adapter class to adapt the log levels for different log libs.
+class SeverityMapper {
+ public:
+  inline static int GetMappedSeverity(int severity) {
+    // Ray log level starts from -1 (RAY_DEBUG);
+    return severity_map[severity + 1];
+  }
+
+ private:
+  static int severity_map[];
+};
+
+void *RayLog::static_impl = nullptr;
+int RayLog::severity_threshold_ = RAY_ERROR;
+
+#ifdef RAY_USE_GLOG
+using namespace google;
+
+// Glog's severity map.
+int SeverityMapper::severity_map[] = {
```

Comment:
Let's make this a function instead of a variable. In that case it can't be overwritten and would be thread safe.